book.md

```diff
@@ -177,7 +177,146 @@ vulnerabilities, along with their mitigations, presented in a rough
+the victim, and parse the victim's responses. These functions can be easily
+chained together, interleaved with arbitrary computations, all written in C, to
+form the exploit.
+
+It's interesting to note that while the read/write primitives consist of
+carefully constructed inputs to the victim program, the attacker can view these
+inputs as *instructions* to the victim program). This duality is
+explored in [@Dullien2020].
+
+In the case of a non-interactive attack, all computation happens within the
+victim program. The duality of input data and code is even more obvious in this
+case, as the malicious input to the victim can be viewed as the exploit code.
+There are cases for which the input is obviously interpreted as code by the
+victim application as well, as in the case of a JavaScript program given as
+input to a JavaScript engine. In this case, the read/write primitives would
+be written as JavaScript functions, which when called have the unintended
+side-effect of accessing arbitrary memory that a JavaScript program is not
+supposed to have access to.  The primitives can be chained together with
+arbitrary computations, also expressed in JavaScript.
+
+There are, however, cases, where the correspondence between data and code isn't
+as obvious. For example, in [@Beer2021], the malicious input consists of a PDF
+file, masquerading as a GIF. Due to an integer overflow bug in the PDF decoder,
+the malicious input leads to an unbounded buffer access.  In the case of
+JavaScript engine exploitation, the attacker would normally be able to use
+JavaScript operations and perform arbitrary computations, making exploitation
+more straightforward. In this case, there are no scripting capabilities
+officially supported. The attackers, however, take advantage of the compression
+format intricacies to implement a small computer architecture, in thousands of
+simple commands to the decoder.  In this way, they effectively _introduce_
+scripting capabilities and are able to express their exploit as a program to
+this architecture.
+
+We have described read/write primitives, and have discussed how an attacker
+might also perform arbitrary computations, either in an external program
+in the case of interactive attacks, or by using scripting capabilities
+(whether originally supported or introduced by the attacker) in
+non-interactive attacks. Assuming an attacker has gained these capabilities,
+how can they use them to achieve their goals?
+
+The ultimate goal of an attacker may vary: it may be, among other things,
+getting access to a system, leaking sensitive information or bringing down a
+service. Frequently, a first step towards these wider goals is arbitrary code
+execution within the victim process. We have already mentioned that the
+attacker will typically have arbitrary computation capabilities at this point,
+but arbitrary code execution also involves things like calling arbitrary
+library functions, performing system calls, and so on.
```

Comment:
Having both "things like" and "and so on" is redundant, and could possibly be simplified to just one of these.

Reply:
Indeed, changing this.